hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFilePreadReader.java

```diff
@@ -0,0 +1,213 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.io.hfile;
+
+import java.io.IOException;
+import java.security.Key;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.io.crypto.Cipher;
+import org.apache.hadoop.hbase.io.crypto.Encryption;
+import org.apache.hadoop.hbase.io.hfile.HFile.FileInfo;
+import org.apache.hadoop.hbase.io.hfile.HFile.Reader;
+import org.apache.hadoop.hbase.security.EncryptionUtil;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link HFile.Reader} to deal with pread.
+ */
+@InterfaceAudience.Private
+public class HFilePreadReader extends HFileReaderImpl {
+  private static final Logger LOG = LoggerFactory.getLogger(HFileReaderImpl.class);
+
+  /**
+   * The iterator will track all blocks in load-on-open section, since we use the
+   * {@link org.apache.hadoop.hbase.io.ByteBuffAllocator} to manage the ByteBuffers in block now,
+   * so we must ensure that deallocate all ByteBuffers in the end.
+   */
+  private final HFileBlock.BlockIterator blockIter;
+
+  public HFilePreadReader(ReaderContext context, FixedFileTrailer trailer,
+      CacheConfig cacheConf, Configuration conf) throws IOException {
+    super(context, cacheConf, conf);
+    this.trailer = trailer;
+    this.compressAlgo = trailer.getCompressionCodec();
+    // Comparator class name is stored in the trailer in version 2.
+    this.comparator = trailer.createComparator();
+    checkFileVersion();
+    this.hfileContext = createHFileContext(context, trailer, conf);
+    this.fsBlockReader = new HFileBlock.FSReaderImpl(context, hfileContext,
+        cacheConf.getByteBuffAllocator());
+    // Initialize an block iterator, and parse load-on-open blocks in the following.
+    this.blockIter = fsBlockReader.blockRange(trailer.getLoadOnOpenDataOffset(),
+      fileSize - trailer.getTrailerSize());
+
+    dataBlockIndexReader = new HFileBlockIndex.CellBasedKeyBlockIndexReader(comparator,
+        trailer.getNumDataIndexLevels());
+    metaBlockIndexReader = new HFileBlockIndex.ByteArrayKeyBlockIndexReader(1);
+    // Data index. We also read statistics about the block index written after
+    // the root level.
+    dataBlockIndexReader.readMultiLevelIndexRoot(
+        blockIter.nextBlockWithBlockType(BlockType.ROOT_INDEX),
+        trailer.getDataIndexCount());
+    // Meta index.
+    metaBlockIndexReader.readRootIndex(
+        blockIter.nextBlockWithBlockType(BlockType.ROOT_INDEX),
+        trailer.getMetaIndexCount());
+    loadFileInfo();
```

Comment:
Maybe we should move the below code out to another place, maybe call it initializeHFileMetaAndIndex? And after done, we use the FSDataInputStream to create a common HFilePreadReader which can be shared by all pread requests? This will be cleaner.

Reply:
Move the code to HFileInfo